File: snippets/go/load-typed-environment-settings-with-go/load_typed_env_settings.go
package main

import (
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

type AppSettings struct {
	AppEnv string
	Port   int
	Debug  bool
}

func LoadTypedEnvSettings(env map[string]string) (AppSettings, error) {
	// Flow:
	//   read APP_ENV, PORT, and DEBUG from the provided map or OS environment
	//      |
	//      +-> parse text, int, and bool values -> return AppSettings
	//      `-> missing or invalid values -> return an error
	appEnv, err := readRequiredText(env, "APP_ENV")
	if err != nil {
		return AppSettings{}, err
	}

	port, err := readInt(env, "PORT", 8000)
	if err != nil {
		return AppSettings{}, err
	}

	debug, err := readBool(env, "DEBUG", false)
	if err != nil {
		return AppSettings{}, err
	}

	return AppSettings{AppEnv: appEnv, Port: port, Debug: debug}, nil
}

func readRequiredText(env map[string]string, key string) (string, error) {
	value, ok := lookupValue(env, key)
	if !ok || strings.TrimSpace(value) == "" {
		return "", fmt.Errorf("%s is required", key)
	}
	return strings.TrimSpace(value), nil
}

func readInt(env map[string]string, key string, defaultValue int) (int, error) {
	value, ok := lookupValue(env, key)
	if !ok {
		return defaultValue, nil
	}

	parsedValue, err := strconv.Atoi(strings.TrimSpace(value))
	if err != nil {
		return 0, fmt.Errorf("%s must be a valid integer", key)
	}
	return parsedValue, nil
}

func readBool(env map[string]string, key string, defaultValue bool) (bool, error) {
	value, ok := lookupValue(env, key)
	if !ok {
		return defaultValue, nil
	}

	normalizedValue := strings.ToLower(strings.TrimSpace(value))
	if normalizedValue == "1" || normalizedValue == "true" {
		return true, nil
	}
	if normalizedValue == "0" || normalizedValue == "false" {
		return false, nil
	}

	return false, fmt.Errorf("%s must be one of: true, false, 1, 0", key)
}

func lookupValue(env map[string]string, key string) (string, bool) {
	if env != nil {
		value, ok := env[key]
		return value, ok
	}
	return os.LookupEnv(key)
}

func main() {
	settings, err := LoadTypedEnvSettings(map[string]string{
		"APP_ENV": "development",
		"PORT":    "8000",
		"DEBUG":   "false",
	})
	if err != nil {
		log.Fatal(err)
	}

	log.Printf("settings=%+v", settings)
}

File: snippets/python/load-typed-env-settings-with-python/src/load_typed_env_settings.py
import os
from dataclasses import dataclass
from typing import Mapping, Optional


@dataclass(frozen=True)
class AppSettings:
    app_env: str
    port: int
    debug: bool


def load_typed_env_settings(env: Optional[Mapping[str, str]] = None) -> AppSettings:
    """Load a small typed settings object from environment variables."""
    # Flow:
    #   read APP_ENV, PORT, and DEBUG from the source mapping
    #      |
    #      +-> parse text, int, and bool values -> return AppSettings
    #      `-> missing or invalid values -> raise ValueError
    source = os.environ if env is None else env

    app_env = _read_required_text(source, "APP_ENV")
    port = _read_int(source, "PORT", default=8000)
    debug = _read_bool(source, "DEBUG", default=False)

    return AppSettings(app_env=app_env, port=port, debug=debug)


def _read_required_text(source: Mapping[str, str], key: str) -> str:
    value = source.get(key)
    if not isinstance(value, str) or value.strip() == "":
        raise ValueError(f"{key} is required")
    return value.strip()


def _read_int(source: Mapping[str, str], key: str, default: int) -> int:
    value = source.get(key)
    if value is None:
        return default
    if not isinstance(value, str):
        raise ValueError(f"{key} must be a string integer")
    try:
        return int(value.strip())
    except ValueError as error:
        raise ValueError(f"{key} must be a valid integer") from error


def _read_bool(source: Mapping[str, str], key: str, default: bool) -> bool:
    value = source.get(key)
    if value is None:
        return default
    if not isinstance(value, str):
        raise ValueError(f"{key} must be a string boolean")

    normalized = value.strip().lower()
    if normalized in {"1", "true"}:
        return True
    if normalized in {"0", "false"}:
        return False

    raise ValueError(f"{key} must be one of: true, false, 1, 0")


if __name__ == "__main__":
    settings = load_typed_env_settings(
        {
            "APP_ENV": "development",
            "PORT": "8000",
            "DEBUG": "false",
        }
    )
    print(settings)